lib/upload.js

```diff
@@ -7,25 +7,71 @@ var path = require('path')
 
 var utils = require('./utils')
 
+var createDirectory = function (container, path, directoryName) {
+  var opts = {
+    name: directoryName,
+    path: path,
+    isDir: true
+  }
+  return Promise.fromCallback(function (callback) {
+    container.createFile(opts, callback)
+  })
+}
+
+var createFile = function (container, path, fileName, file) {
+  var opts = {
+    name: fileName,
+    path: path,
+    isDir: false,
+    content: file.toString()
+  }
+  return Promise.fromCallback(function (callback) {
+    container.createFile(opts, callback)
+  })
+}
+
+var recusivelyCreateDirectories = function (container, basePath, fullPath) {
+  function createSingleDirectory (createdPaths, remainingPaths) {
+    if (remainingPaths.length > 0) {
+      var createPath = path.join(basePath, createdPaths.join(path.sep))
+      var newPath = remainingPaths.splice(0, 1)[0]
+      createdPaths.push(newPath)
+      return createDirectory(container, createPath, newPath)
+        .then(function () {
+          return createSingleDirectory(createdPaths, remainingPaths)
+        })
+        .catch(function () {
+          return createSingleDirectory(createdPaths, remainingPaths)
+        })
+    }
+    return Promise.resolve()
+  }
+  var allPaths = fullPath.split(path.sep)
+  if (allPaths[allPaths.length - 1].length === 0) {
+    allPaths.pop()
+  }
+  return createSingleDirectory([], allPaths)
+}
```

Comment:
I dislike in-file functions. these are impossible to have written any individual test for (if you needed to). Move these to the `module.exports` with `_` prefixes - you can assign `module.exports` to a variable to make it easier as well (like (login)[https://github.com/Runnable/cli/blob/SAN-3573-upload-command/lib/login.js])